react/more-react-concepts/managing-state-with-context-api/managing-state-with-context-api.md

```diff
@@ -0,0 +1,259 @@
+    <div>
+      <img />
+      <div>
+        {/* elements that aligns with the design */}
+        <button type="button" onClick={() => addToCart(product)}>
+          Add to Cart
+        </button>
+      </div>
+    </div>
+  );
+}
+~~~
+
+Great. This is a very common pattern, we pass data/functions between components, nothing too complicated. But there's a lot to micro-manage, a lot of components that you need to pass props to, and there's even a little bit of prop drilling that's happening, where the `cartItemsCount` is passed from `App` -> `Header` -> `Links`.
+
+Let's elaborate everything that's happening
+
+- `products` and `addToCart` is passed down to the `Product` component
+- `cartItemsCount` is passed down to the `Header` and `Links` components
+
+Very simple application but imagine the application grows in size, with more features added say in the Product Detail Page, a lot more components will be nested inside the component and potentially more nested props. What if we have a `Cart` component, a `ProductListing` component, what if we added more functionalities other than just `addToCart`. It can get repetitive and complex because we all know, more features equals more complexities.
+
+### Implementing Context API
+
+Let's try to implement Context API and refactor our current application to reduce these complexities.
+
+We have three things in the API that we need use in order to implement it.
+
+1. `createContext` - This "creates the context" duh... But yes, it's how we can create the context, it takes in any value be it a number, string, or an object, which can be referred to as the _default state_ of the context, and returns a Context object
+2. `useContext` - This is a hook that takes in the Context object that is returned by `createContext`, we will be using this inside our component in order to "get" the data that we need
+3. `ContextObject.Provider` - The Context object comes with the `Provider` component that has a prop called `value` which is basically the context value that's going to be passed down to the components no matter how deep they're nested. In other words, a way to "provide" the context value to these components.
+
+We can start by using the `createContext` function that can be imported from the `react` module.
+
+~~~javascript
+import { createContext } from "react";
+~~~
+
+The createContext is able to take in any value, a string, number, or even an object, which can be referred to as the _default state_ of the context.
+
+In our case, we'll use an object that contains the following shape:
+
+~~~javascript
+const ShopContext = createContext({
+  products: [],
+  cartItems: [],
+  addToCart: () => {},
+});
+~~~
+
+You might think, adding `products` and `cartItems` seems reasonble, but why do we need to include `addToCart` as well? Actually, you're not required to setup your your context like this and it's up to you, but you will be able to add auto-completion in your IDE by doing this, it's basically a bonus!
```

Comment:
> Actually, you're not required to setup your your context like this and it's up to you, but you will be able to add auto-completion in your IDE by doing this, it's basically a bonus!

I'm not sure how adding `addToCart` field enables auto-completion in IDE. I'll try it myself in a while. That said, can you elaborate on this please.

Reply:
Without:

![image](https://user-images.githubusercontent.com/75290989/229119727-110fa3d1-1219-4d07-b4f3-79c827a46c9d.png)
![image](https://user-images.githubusercontent.com/75290989/229119469-fbc68a5d-5303-4769-9e94-0c9e3cc26d42.png)

With:
![image](https://user-images.githubusercontent.com/75290989/229119651-119a5c38-8be4-40e3-88ea-84aa78e2dc4e.png)
![image](https://user-images.githubusercontent.com/75290989/229119685-d1f64070-bd6c-453c-9e06-eea612f3df3a.png)